File: apps/web/src/shared/api/discoveredTools.ts
/**
 * Discovered Tools API client
 */
import { apiRequest } from './http';

export interface DiscoveredToolListItem {
  id: string;
  tool_id?: string | null;
  slug: string;
  name: string;
  description: string | null;
  source: string;
  provider_instance_id: string | null;
  connector_slug: string | null;
  connector_name: string | null;
  domains: string[];
  is_active: boolean;
  last_seen_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface DiscoveredToolDetail extends DiscoveredToolListItem {
  input_schema: Record<string, unknown> | null;
  output_schema: Record<string, unknown> | null;
}

export interface RescanResponse {
  message: string;
  stats: Record<string, unknown>;
}

export interface DiscoveredToolUpdateRequest {
  tool_id?: string | null;
}

export interface McpProbeToolItem {
  slug: string;
  name: string;
  description?: string | null;
  has_input_schema: boolean;
  has_output_schema: boolean;
}

export interface McpProbeResponse {
  provider_instance_id: string;
  provider_slug: string;
  provider_url: string;
  tools_count: number;
  tools: McpProbeToolItem[];
}

export interface McpOnboardRequest {
  provider_instance_id: string;
  enable_all_in_runtime?: boolean;
  include_local?: boolean;
}

export interface McpOnboardResponse {
  provider_instance_id: string;
  probe_tools_count: number;
  rescan_stats: Record<string, unknown>;
  linked_tools_updated: number;
  active_discovered_tools: number;
  published_tools: number;
}

export const discoveredToolsApi = {
  list: async (params?: {
    source?: string;
    domain?: string;
    is_active?: boolean;
  }): Promise<DiscoveredToolListItem[]> => {
    const searchParams = new URLSearchParams();
    if (params?.source) searchParams.set('source', params.source);
    if (params?.domain) searchParams.set('domain', params.domain);
    if (params?.is_active !== undefined) searchParams.set('is_active', String(params.is_active));
    const qs = searchParams.toString();
    return apiRequest<DiscoveredToolListItem[]>(
      `/admin/discovered-tools${qs ? `?${qs}` : ''}`,
      { method: 'GET' },
    );
  },

  get: async (id: string): Promise<DiscoveredToolDetail> => {
    return apiRequest<DiscoveredToolDetail>(`/admin/discovered-tools/${id}`, { method: 'GET' });
  },

  probeMcp: async (providerInstanceId: string): Promise<McpProbeResponse> => {
    const params = new URLSearchParams({ provider_instance_id: providerInstanceId });
    return apiRequest<McpProbeResponse>(`/admin/discovered-tools/probe-mcp?${params.toString()}`, {
      method: 'POST',
    });
  },

  onboardMcp: async (data: McpOnboardRequest): Promise<McpOnboardResponse> => {
    return apiRequest<McpOnboardResponse>('/admin/discovered-tools/onboard-mcp', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  rescan: async (data: { include_local?: boolean; provider_instance_id?: string | null } = { include_local: true }): Promise<RescanResponse> => {
    return apiRequest<RescanResponse>('/admin/discovered-tools/rescan', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },
};

File: apps/api/src/app/schemas/discovered_tools.py
"""
Schemas for DiscoveredTool admin API.
"""
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class DiscoveredToolListItem(BaseModel):
    id: UUID
    tool_id: Optional[UUID] = None
    slug: str
    name: str
    description: Optional[str] = None
    source: str
    provider_instance_id: Optional[UUID] = None
    connector_slug: Optional[str] = None
    connector_name: Optional[str] = None
    domains: List[str] = Field(default_factory=list)
    is_active: bool = True
    last_seen_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DiscoveredToolDetailResponse(BaseModel):
    id: UUID
    tool_id: Optional[UUID] = None
    slug: str
    name: str
    description: Optional[str] = None
    source: str
    provider_instance_id: Optional[UUID] = None
    connector_slug: Optional[str] = None
    connector_name: Optional[str] = None
    domains: List[str] = Field(default_factory=list)
    input_schema: Optional[Dict[str, Any]] = None
    output_schema: Optional[Dict[str, Any]] = None
    is_active: bool = True
    last_seen_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RescanResponse(BaseModel):
    message: str
    stats: Dict[str, Any]


class DiscoveredToolsRescanRequest(BaseModel):
    include_local: bool = True
    provider_instance_id: Optional[UUID] = None


class McpProbeToolItem(BaseModel):
    slug: str
    name: str
    description: Optional[str] = None
    has_input_schema: bool
    has_output_schema: bool


class McpProbeResponse(BaseModel):
    provider_instance_id: UUID
    provider_slug: str
    provider_url: str
    tools_count: int
    tools: List[McpProbeToolItem] = Field(default_factory=list)


class McpOnboardRequest(BaseModel):
    provider_instance_id: UUID
    enable_all_in_runtime: bool = False
    include_local: bool = False


class McpOnboardResponse(BaseModel):
    provider_instance_id: UUID
    probe_tools_count: int
    rescan_stats: Dict[str, Any]
    linked_tools_updated: int
    active_discovered_tools: int
    published_tools: int


class DiscoveredToolUpdateRequest(BaseModel):
    tool_id: Optional[UUID] = Field(
        default=None,
        description="Publication container link. NULL means discovered-only capability.",
    )
